weave/feedback.py

```diff
@@ -0,0 +1,230 @@
+        feedback = item
+
+        typ = feedback.feedback_type
+        display_type = typ
+        if typ == "wandb.reaction.1":
+            display_type = "reaction"
+            if util.is_notebook():
+                # TODO: Emojis mess up table alignment in Jupyter 😢
+                #       maybe there is something else we could do here?
+                content = feedback.payload["alias"]
+            else:
+                content = feedback.payload["emoji"]
+        elif typ == "wandb.note.1":
+            display_type = "note"
+            content = feedback.payload["note"]
+        else:
+            content = json.dumps(feedback.payload, indent=2)
+
+        # TODO: Prettier relative time display?
+        created_at = str(feedback.created_at.replace(tzinfo=None))
+
+        creator = feedback.wb_user_id
+        if feedback.creator is not None:
+            creator = f"{feedback.creator} ({creator})"
+
+        row = [
+            display_type,
+            content,
+            created_at,
+            feedback.id,
+            creator,
+        ]
+        if self.show_refs:
+            row.insert(0, feedback.weave_ref)
+        return row
+
+
+class FeedbackQuery:
+    """Lazy-loading object for fetching feedback from the server."""
+
+    entity: str
+    project: str
+
+    show_refs: bool
+    _query: tsi.Query
+    offset: Optional[int]
+    limit: Optional[int]
+
+    feedbacks: Optional[Feedbacks]
+
+    def __init__(
+        self,
+        entity: str,
+        project: str,
+        query: Query,
+        offset: Optional[int] = None,
+        limit: Optional[int] = None,
+        show_refs: bool = False,
+    ):
+        self.client = graph_client_context.require_graph_client()
+        self.entity = entity
+        self.project = project
+
+        self.show_refs = show_refs
+        self._query = query
+        self.offset = offset
+        self.limit = limit
+
+        self.feedbacks = None
+
+    def __iter__(self) -> Iterator[tsi.Feedback]:
+        yield from self.execute()
+
+    def refresh(self) -> Feedbacks:
+        sort_by = [
+            {
+                "field": "created_at",
+                "direction": "asc",
+            },
+        ]
+        req = tsi.FeedbackQueryReq(
+            project_id=f"{self.entity}/{self.project}",
+            query=self._query,
+            sort_by=sort_by,
+            offset=self.offset,
+            limit=self.limit,
+        )
+        response = self.client.server.feedback_query(req)
+        # Response is dicts because API allows user to specify fields, but we don't
+        # expose that in this Python API.
+        return Feedbacks(self.show_refs, (tsi.Feedback(**r) for r in response.result))
+
+    def execute(self) -> Feedbacks:
+        if self.feedbacks is not None:
+            return self.feedbacks
+        self.feedbacks = self.refresh()
+        return self.feedbacks
+
+    def refs(self) -> Refs:
+        return self.execute().refs()
```

Comment:
i wonder if this should be doing some sort of unique operation over the refs?

Reply:
It does.

Reply:
oh nice, missed this: `uris = list(set(feedback.weave_ref for feedback in self.items))`

Reply:
hmmm well now that i think about this, doesn't that mean that ordering is not guaranteed anymore? Which would be unexpected if the order changes for each attempt to get calls for a certain reaction

Reply:
Good catch! For some reason I was under the mistaken belief that they made sets preserve insertion order in 3.7 when they guaranteed it for dicts. Will fix it to use dict.fromkeys.

Reply:
As my kids are fond of saying, "this is sus". :-) 
https://github.com/wandb/weave/blob/jamie/feedback-python-api/weave/tests/test_client_trace.py#L378